### ISSUE TITLE
Queue submit and event signal ordering clarification

### ISSUE THREAD

**Issue by nanokatze** [KhronosGroup/Vulkan-Docs#1464]
Section 6.6. Queue Forward Progress says:

> Command buffers in the submission can include vkCmdWaitEvents commands that wait on events that will not be signaled by earlier commands in the queue. Such events must be signaled by the application using vkSetEvent, and the vkCmdWaitEvents commands that wait upon them must not be inside a render pass instance. **The event must be set before the vkCmdWaitEvents command is executed.**

Assume the queue is not waiting on any host-signaled semaphores. This statement suggests that a queue submit followed by an event signal would not be allowed, as in such case there is no way to ensure that event signal happens before event wait. Yet one of the VU items in vkCmdWaitEvents explicitly calls out this case:

> If pEvents includes one or more events that will be signaled by vkSetEvent after commandBuffer has been submitted to a queue, then vkCmdWaitEvents must not be called inside a render pass instance

It is thus unclear, is it okay to submit a command buffer that waits for event A followed by signal of A?

**Comment by krOoze** [comment]
Previously it was allowed, now it is unspecified behavior.

**Comment by devshgraphicsprogramming** [comment]
it seems that there's literally no use for a CPU signalled event, because the "GPU wait cannot wait on the GPU"

**Comment by Tobski** [comment]
We added CPU event setting because there were some uses for it on some platforms (I believe this was used for late-latching of data in VR use cases at some point), however we couldn't agree consistent enough portable behavior to put anything concrete in the spec. Eventually we conceded that trying to say anything useful in the spec was just confusing developers, so we simply amended it to say "it's undefined behavior". 

If you have a specific use case, and are targetting a specific platform, it might be possible to discuss the exact behaviour with the GPU vendor on that platform, but we can't recommend them for general use. Otherwise yes, event signalling on the host is basically useless.

The VU you quoted remains correct if useless. It's mostly there such that existing niche use cases continue to work within reasonable bounds, and in case we ever figure out how to define a useful portable behaviour.
